src/Castle.Core/DynamicProxy/Generators/GeneratorUtil.cs

```diff
@@ -74,25 +76,23 @@ bool IsReadOnly(ParameterInfo parameter)
 				//
 				// The above points inform the following detection logic: First, we rely on an IL
 				// `[in]` modifier being present. This is a "fast guard" against non-`in` parameters:
-				if ((parameter.Attributes & (ParameterAttributes.In | ParameterAttributes.Out)) == ParameterAttributes.In)
-				{
-					// Here we perform the actual check. We don't rely on cmods because support
-					// for them is at current too unreliable in general, and because we wouldn't
-					// be saving much time anyway.
-					if (parameter.GetCustomAttributes(false).Any(IsIsReadOnlyAttribute))
-					{
-						return true;
-					}
-				}
+				if ((parameter.Attributes & (ParameterAttributes.In | ParameterAttributes.Out)) != ParameterAttributes.In)
+					return false;
 
-				return false;
+#if FEATURE_CUSTOMMODIFIERS
+				// Check for the required modifiers first, as it's faster.
+				// This check allows to make the detection logic more robust on the platforms which support custom modifiers.
+				// If check fails, resort to the IsReadOnlyAttribute check.
+				if (parameter.GetRequiredCustomModifiers().Any(x => x == typeof(InAttribute)))
+					return true;
+#endif
 
-				bool IsIsReadOnlyAttribute(object attribute)
-				{
-					// The comparison by name is intentional; any assembly could define that attribute.
-					// See explanation in comment above.
-					return attribute.GetType().FullName == "System.Runtime.CompilerServices.IsReadOnlyAttribute";
-				}
+				// The comparison by name is intentional; any assembly could define that attribute.
+				// See explanation in comment above.
+				if (parameter.GetCustomAttributes(false).Any(x => x.GetType().FullName == "System.Runtime.CompilerServices.IsReadOnlyAttribute"))
+					return true;
```

Comment:
The changes look good. Only nit, could you wrap the 3 if-statement bodies with curly braces as per our coding standard.